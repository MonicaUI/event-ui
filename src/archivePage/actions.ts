import v1PageMigrations from 'nucleus-widgets/migrations/website/v1/pages';

export const LOAD_EVENT_ARCHIVE_PAGE_DATA = 'event-guestside-site/LOAD_EVENT_ARCHIVE_PAGE_DATA';

export const transformAndLoadEventArchivedPageData = (eventArchivePageData: $TSFixMe) => {
  return (dispatch: $TSFixMe, getState: $TSFixMe): $TSFixMe => {
    const { website: existingWebsite } = getState();
    dispatch({
      type: LOAD_EVENT_ARCHIVE_PAGE_DATA,
      payload: {
        website: {
          ...existingWebsite,
          pages: v1PageMigrations(eventArchivePageData.pages),
          layoutItems: eventArchivePageData.layoutItems,
          pluginData: eventArchivePageData.plugins
        }
      }
    });
  };
};
