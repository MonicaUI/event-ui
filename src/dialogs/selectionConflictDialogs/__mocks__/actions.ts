export const returnToProcessStart = jest.fn(() => () => {});
