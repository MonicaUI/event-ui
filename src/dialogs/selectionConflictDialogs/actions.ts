import { closeDialogContainer } from 'nucleus-guestside-site/src/redux/modules/dialogContainer';
import { routeToPage } from '../../redux/pathInfo';
import { REGISTRATION } from '../../redux/website/registrationProcesses';

export function returnToProcessStart() {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    dispatch(closeDialogContainer());
    await dispatch(routeToPage(REGISTRATION.forCurrentRegistrant().startPageId(getState())));
  };
}
