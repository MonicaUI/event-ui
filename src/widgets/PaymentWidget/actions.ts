/* Method for adding / removing taxes in Order Summary for planner reg */
import { getRegCart } from '../../redux/selectors/shared';
import {
  DISCOUNT_CODE_VALIDATION_FAILURE,
  UPDATE_REG_CART_FAILURE,
  UPDATE_REG_CART_PENDING,
  UPDATE_REG_CART_SUCCESS
} from '../../redux/registrationForm/regCart/actionTypes';
import { closeDialogContainer, hideLoadingOnError } from 'nucleus-guestside-site/src/redux/modules/dialogContainer';
import { unset, updateIn } from 'icepick';
import {
  getAllowableEventRegistrationsForDiscountCapacityMap,
  getAvailableCapacity,
  getRegistrantsCountedTowardsDiscountCapacity,
  getRegistrantsEligibleForDiscount,
  getRegistrationTypeId,
  getRegistrationPathId
} from '../../redux/registrationForm/regCart/selectors';
import { findKnownErrorResourceKey, getUpdateErrors } from '../../redux/registrationForm/errors';
import { loadAvailableCapacityCounts } from '../../redux/capacity';
import { getEventRegistrationId, isGroupRegistration } from '../../redux/selectors/currentRegistrant';
import { openGuestProductSelectionDialog, openAlreadyRegisteredDialog } from '../../dialogs';
import Logger from '@cvent/nucleus-logging';
import { updateRegCart } from '../../redux/registrationForm/regCart';

const LOG = new Logger('widgets/PaymentWidget/actions');

export function updateIgnoreTaxesInRegCart(ignoreTaxes: $TSFixMe) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const state = getState();
    const {
      accessToken,
      clients: { regCartClient },
      event,
      account
    } = state;
    const regCart = getRegCart(state);
    let response;
    dispatch({ type: UPDATE_REG_CART_PENDING });
    try {
      response = await updateRegCart(state, regCartClient, accessToken, {
        ...regCart,
        ignoreTaxes
      });
      dispatch({
        type: UPDATE_REG_CART_SUCCESS,
        payload: {
          regCart: response.regCart,
          validationMessages: response.validationMessages
        }
      });
      LOG.debug(`updateRegCart with id ${regCart.regCartId} was successful`);
    } catch (error) {
      // if we get external auth or oauth error , we need to redirect to external auth or oauth url
      if (
        getUpdateErrors.handleAuthError(
          error,
          account,
          event,
          getRegistrationTypeId(regCart, getEventRegistrationId(getState())),
          getRegistrationPathId(regCart, getEventRegistrationId(getState()))
        )
      ) {
        return;
      }
      dispatch({ type: UPDATE_REG_CART_FAILURE, payload: { error } });
      dispatch(hideLoadingOnError());

      throw error;
    }
  };
}

export function updateIgnoreServiceFeesInRegCart(ignoreServiceFee: $TSFixMe) {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const state = getState();
    const {
      accessToken,
      clients: { regCartClient },
      event,
      account
    } = state;
    const regCart = getRegCart(state);
    let response;
    dispatch({ type: UPDATE_REG_CART_PENDING });
    try {
      response = await updateRegCart(state, regCartClient, accessToken, {
        ...regCart,
        ignoreServiceFee
      });
      dispatch({
        type: UPDATE_REG_CART_SUCCESS,
        payload: {
          regCart: response.regCart,
          validationMessages: response.validationMessages
        }
      });
      LOG.debug(`updateRegCart with id ${regCart.regCartId} was successful`);
    } catch (error) {
      // if we get external auth or oauth error , we need to redirect to external auth or oauth url
      if (
        getUpdateErrors.handleAuthError(
          error,
          account,
          event,
          getRegistrationTypeId(regCart, getEventRegistrationId(getState())),
          getRegistrationPathId(regCart, getEventRegistrationId(getState()))
        )
      ) {
        return;
      }
      dispatch({ type: UPDATE_REG_CART_FAILURE, payload: { error } });
      dispatch(hideLoadingOnError());

      throw error;
    }
  };
}

/**
 * applies a discount code to the reg cart
 */
export function updateDiscountCodes(
  discountCode: $TSFixMe,
  removeDiscount: $TSFixMe,
  eventRegistrationsToApplyCodeTo?: $TSFixMe,
  previousAvailableCapacity?: $TSFixMe,
  previousErrors?: $TSFixMe
) {
  // eslint-disable-next-line complexity
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      accessToken,
      regCartStatus: { lastSavedRegCart },
      clients: { regCartClient },
      event,
      account
    } = getState();
    dispatch({ type: UPDATE_REG_CART_PENDING });
    const regCart = getRegCart(getState());

    /*
     * when we try to apply new discount we need to use the latest reg cart, so we can calculate the
     * discount based on current reg cart info, this is needed for discount advance filter
     */
    let updatedRegCart = {};
    if (removeDiscount) {
      updatedRegCart = updateIn(lastSavedRegCart, ['discounts'], discountCodeMap =>
        unset(discountCodeMap, discountCode.toLowerCase())
      );
    } else if (eventRegistrationsToApplyCodeTo) {
      updatedRegCart = updateIn(regCart, ['discounts'], discountCodeMap => {
        return { ...discountCodeMap, [discountCode]: { discountCode, eventRegistrationsToApplyCodeTo } };
      });
    } else {
      updatedRegCart = updateIn(regCart, ['discounts'], discountCodeMap => {
        return { ...discountCodeMap, [discountCode]: { discountCode } };
      });
    }
    try {
      const response = await updateRegCart(getState(), regCartClient, accessToken, updatedRegCart);
      dispatch({
        type: UPDATE_REG_CART_SUCCESS,
        payload: {
          regCart: response.regCart,
          validationMessages: response.validationMessages
        }
      });

      LOG.debug(`updateRegCart with id ${lastSavedRegCart.regCartId} was successful`);
    } catch (error) {
      // if we get external auth or oauth error , we need to redirect to external auth or oauth url
      if (
        getUpdateErrors.handleAuthError(
          error,
          account,
          event,
          getRegistrationTypeId(updatedRegCart, getEventRegistrationId(getState())),
          getRegistrationPathId(updatedRegCart, getEventRegistrationId(getState()))
        )
      ) {
        return;
      }
      dispatch({ type: UPDATE_REG_CART_FAILURE, payload: { error } });
      dispatch(hideLoadingOnError());

      let validationError = null;

      const state = getState();
      // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
      const errors = state.registrationForm && state.registrationForm.errors;
      const currentAvailableCapacity = getAvailableCapacity(errors);

      if (getUpdateErrors.isDiscountCapacityInsufficient(error)) {
        if (previousAvailableCapacity && previousAvailableCapacity !== currentAvailableCapacity) {
          await dispatch(selectDiscountForRegistrant(discountCode, previousErrors));
        } else {
          await dispatch(selectDiscountForRegistrant(discountCode, null));
        }
        return;
      } else if (getUpdateErrors.isProductAvailabilityError(error)) {
        // TODO: display error specific for discount capacity
        dispatch(loadAvailableCapacityCounts());
        validationError = 'EventWidgets_Validations_DiscountCodeCapacityFull__resx';
      } else if (getUpdateErrors.isKnownError(error)) {
        validationError = findKnownErrorResourceKey(error.responseBody.validationMessages);
      }

      if (validationError) {
        return dispatch({
          type: DISCOUNT_CODE_VALIDATION_FAILURE,
          payload: {
            validationError
          }
        });
      }

      throw error;
    }
  };
}

export function updateRegCartForAutoDiscounts() {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      accessToken,
      clients: { regCartClient },
      event,
      account
    } = getState();
    const regCart = getRegCart(getState());
    dispatch({ type: UPDATE_REG_CART_PENDING });
    try {
      const response = await updateRegCart(getState(), regCartClient, accessToken, regCart);
      dispatch({
        type: UPDATE_REG_CART_SUCCESS,
        payload: {
          regCart: response.regCart,
          validationMessages: response.validationMessages
        }
      });
      LOG.debug(`updateRegCart with id ${regCart.regCartId} was successful`);
    } catch (error) {
      if (getUpdateErrors.isInviteeAlreadyRegistered(error)) {
        LOG.debug(`updateRegCart with id ${regCart.regCartId} failed`, error);
        dispatch({ type: UPDATE_REG_CART_FAILURE, payload: { error } });
        dispatch(hideLoadingOnError());
        return dispatch(
          openAlreadyRegisteredDialog({
            title: 'EventGuestSide_ExistingInvitee_AlreadyRegistered_Header__resx',
            instructionalText: 'EventGuestSide_ExistingInvitee_AlreadyRegistered_InstructionalText__resx',
            prepopulateForm: true
          })
        );
      }
      // if we get external auth or oauth error , we need to redirect to external auth or oauth url
      if (
        getUpdateErrors.handleAuthError(
          error,
          account,
          event,
          getRegistrationTypeId(regCart, getEventRegistrationId(getState())),
          getRegistrationPathId(regCart, getEventRegistrationId(getState()))
        )
      ) {
        return;
      }
      dispatch({ type: UPDATE_REG_CART_FAILURE, payload: { error } });
      dispatch(hideLoadingOnError());

      throw error;
    }
  };
}

/**
 * Apply discount to selected registrants only
 */
function selectDiscountForRegistrant(discountCodeValue, previousErrors) {
  return async (dispatch, getState) => {
    const {
      registrationForm,
      regCartStatus: { lastSavedRegCart }
    } = getState();
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const errors = previousErrors || (registrationForm && registrationForm.errors);
    const allowableEventRegistrationsForDiscountCapacityMap =
      getAllowableEventRegistrationsForDiscountCapacityMap(errors);
    // Get only those registrants who count towards discount capacity
    const discountCapacityRegistrants = getRegistrantsCountedTowardsDiscountCapacity(
      lastSavedRegCart,
      allowableEventRegistrationsForDiscountCapacityMap
    );
    dispatch(handleDiscountSelectionForInsufficientCapacity(discountCodeValue, discountCapacityRegistrants));
  };
}

/* Apply discount to registrants selected in the modal */
const applyDiscountToSelectedRegistrants = (
  discountCodeValue,
  currentPrimaryRegId,
  selectedEventRegIds,
  eventRegistrations
) => {
  return async (dispatch, getState) => {
    dispatch(closeDialogContainer());
    const {
      registrationForm,
      regCartStatus: { lastSavedRegCart }
    } = getState();
    const selectedDiscountEligibleRegistrants = [];
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const errors = registrationForm && registrationForm.errors;
    const allowableEventRegistrationsForDiscountCapacityMap =
      getAllowableEventRegistrationsForDiscountCapacityMap(errors);
    const availableCapacity = getAvailableCapacity(errors);
    // Get all registrants who are eligible for the discount
    let discountEligibleRegistrants = [];
    eventRegistrations.forEach(eventReg => {
      const eventRegId = eventReg.eventRegistrationId;
      if (selectedEventRegIds[eventRegId].isSelected) {
        // If this registrant is selected, apply discount code as applicable
        discountEligibleRegistrants = getRegistrantsEligibleForDiscount(
          lastSavedRegCart,
          allowableEventRegistrationsForDiscountCapacityMap,
          eventRegId
        );
        discountEligibleRegistrants.forEach(regId => {
          selectedDiscountEligibleRegistrants.push(regId);
        });
      }
    });
    dispatch(
      updateDiscountCodes(discountCodeValue, false, selectedDiscountEligibleRegistrants, availableCapacity, errors)
    );
  };
};

function handleDiscountSelectionForInsufficientCapacity(discountCodeValue, discountCapacityRegistrants) {
  return async (dispatch, getState) => {
    const state = getState();
    const eventRegistrations = discountCapacityRegistrants;
    /*
     * this variable is needed for the dialog logic, but since we don't need it for the discount logic,
     * we pass '' and check for it in the dialog logic
     */
    const eventRegistrationId = '';
    // Passing discountCodeValue into modal's productId field
    const discountId = discountCodeValue;
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    const errors = state.registrationForm && state.registrationForm.errors;
    const availableCapacity = getAvailableCapacity(errors);
    const discountTitle = 'GuestProductSelection_DiscountTitle__resx';
    // overrideCapacity will always be false in this case
    const overrideCapacity = false;
    const isGroupReg = isGroupRegistration(state);
    const eventRegSelections = {};
    // the eventRegSelections properties would always be false here
    eventRegistrations.forEach(eventReg => {
      const eventRegId = eventReg.eventRegistrationId;
      eventRegSelections[eventRegId] = {
        isSelected: false,
        isDisabled: false,
        registeredForProductInGroup: false
      };
    });
    await dispatch(
      openGuestProductSelectionDialog(
        'GuestProductSelection_ApplyDiscount__resx',
        discountId,
        discountTitle,
        availableCapacity,
        overrideCapacity,
        eventRegSelections,
        eventRegistrations,
        eventRegistrationId,
        isGroupReg,
        applyDiscountToSelectedRegistrants
      )
    );
  };
}
