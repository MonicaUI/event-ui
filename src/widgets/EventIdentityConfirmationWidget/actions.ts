import {
  getEventRegistration,
  getVisibleRegistrationTypesByPlannerSelection
} from '../../redux/selectors/currentRegistrant';
import { getRegistrationPageFields } from '../../redux/selectors/currentRegistrationPath';
import { NOT_VISIBLE, READ_ONLY } from 'cvent-question-widgets/lib/DisplayType';
import { isWidgetReviewed } from '../../redux/website/pageContentsWithGraphQL';
import StandardContactFields from 'event-widgets/lib/StandardContactFields/StandardContactFields';
import { defaultRegistrationTypeId } from 'event-widgets/utils/registrationType';
import { CLEAR_REG_CART_INFERRED_FIELDS } from '../../redux/registrationForm/regCart/actionTypes';
import { setIn } from 'icepick';
import { getRegCart } from '../../redux/selectors/shared';
import { saveRegistration } from '../../redux/registrationForm/regCart';
import { runRegistrationTypeChangeValidationsForPrimaryAndGuest } from '../../redux/registrationForm/regCart';
import {
  handleRegistrationTypeSelectionConflict,
  validateUserRegistrationTypeSelection
} from '../../dialogs/selectionConflictDialogs';
import { withLoading } from 'nucleus-guestside-site/src/redux/modules/dialogContainer';
import { getUpdateErrors } from '../../redux/registrationForm/errors';
import { openNoAdmissionItemAvailableForRegistrationTypeDialog, openPrivateEventErrorDialog } from '../../dialogs';
import { merge } from 'lodash';
import { populateVisibleProducts } from '../../redux/visibleProducts';
import { clearOutOfSyncUserSessionFieldsAndUrlQuery } from '../../redux/registrantLogin/actions';
import { ApolloClient } from '@apollo/client';
import type { RootState, AsyncAppThunk } from '../../redux/reducer';

export const isRegTypeReviewed = async (
  state: RootState,
  registrationTypeId: string,
  registrationPathId: string,
  apolloClient: ApolloClient<unknown>
): Promise<boolean> => {
  const widgetTypeName = 'RegistrationType';
  const visibleRegTypes = getVisibleRegistrationTypesByPlannerSelection(state, registrationPathId) || [];
  const isRegTypeWidgetReviewed = await isWidgetReviewed(state, { widgetType: widgetTypeName }, apolloClient);
  return visibleRegTypes.includes(registrationTypeId) && isRegTypeWidgetReviewed;
};

export function clearFields(): AsyncAppThunk {
  return async (
    dispatch,
    getState,
    {
      apolloClient
    }: {
      apolloClient?: ApolloClient<unknown>;
    }
  ) => {
    try {
      const state = getState();
      const eventRegistration = JSON.parse(JSON.stringify(getEventRegistration(state)));
      const isRegTypeWidgetReviewed = await isRegTypeReviewed(
        state,
        eventRegistration.registrationTypeId,
        eventRegistration.registrationPathId,
        apolloClient
      );
      if (state.userSession.regTypeId && !isRegTypeWidgetReviewed) {
        const regTypeChangeValidations = await runRegistrationTypeChangeValidationsForPrimaryAndGuest(
          validateUserRegistrationTypeSelection,
          defaultRegistrationTypeId,
          getState(),
          true,
          apolloClient
        );
        // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
        if (regTypeChangeValidations && regTypeChangeValidations.hasConflicts) {
          return dispatch(
            handleRegistrationTypeSelectionConflict(
              regTypeChangeValidations.validationResults,
              null,
              null,
              dissociateInviteeAndClearInferredFields
            )
          );
        }
      }
      const linkedInviteeIdToRemove = eventRegistration.attendee.attendeeId;
      if (state.userSession.regTypeId && !isRegTypeWidgetReviewed) {
        eventRegistration.registrationTypeId = defaultRegistrationTypeId;
      }
      /*
       * This is an existing pattern that we will need to revisit later. We should clear
       * the individual field rather than replace the entire reg cart
       */
      dispatch({
        type: CLEAR_REG_CART_INFERRED_FIELDS,
        payload: {
          regCart: setIn(
            getRegCart(state),
            ['eventRegistrations', eventRegistration.eventRegistrationId],
            eventRegistration
          )
        }
      });
      /**
       * on clear fields regType is cleared, need to reevaluate visible products
       */
      await dispatch(populateVisibleProducts(eventRegistration.eventRegistrationId));
      await dispatch(dissociateInviteeAndClearInferredFields());
      return await dispatch(withLoading(saveRegistration)(linkedInviteeIdToRemove));
    } catch (error) {
      if (
        // @ts-expect-error ts-migrate(2554) FIXME: Expected 1 arguments, but got 2.
        getUpdateErrors.isPrivateEvent(error, getState()) ||
        // @ts-expect-error ts-migrate(2554) FIXME: Expected 0-1 arguments, but got 2.
        getUpdateErrors.isAttendeeNotAllowedByCustomLogic(error, getState())
      ) {
        return await dispatch(openPrivateEventErrorDialog());
      }
      if (getUpdateErrors.isAdmissionItemsNotAvailableForRegTypeError(error)) {
        return await dispatch(openNoAdmissionItemAvailableForRegistrationTypeDialog());
      }
      throw error;
    }
  };
}

export function dissociateInviteeAndClearInferredFields(): AsyncAppThunk {
  return async (dispatch, getState) => {
    const state = getState();
    const eventRegistration = JSON.parse(JSON.stringify(getEventRegistration(state)));
    // invitee associated attendeeId and contact Id
    delete eventRegistration.attendee.attendeeId;
    delete eventRegistration.attendee.personalInformation.contactId;
    // inferred information to clear: 2. inviteeId related personalInformation
    const existingContactInfo = eventRegistration.attendee.personalInformation;
    const contactInfo = await Promise.all(
      getRegistrationPageFields(state).map(async field => {
        const isFieldReviewed = await isWidgetReviewed(state, { fieldId: field.fieldId });
        return field.display !== NOT_VISIBLE && field.display !== READ_ONLY && isFieldReviewed ? field : null;
      })
    );
    const filteredContactInfo = contactInfo.filter(Boolean);
    const copiedContactInfo = filteredContactInfo.map(field =>
      field.isCustomField
        ? {
            customFields: {
              [field.fieldId]: existingContactInfo.customFields[field.fieldId]
            }
          }
        : {
            // @ts-expect-error ts-migrate(2464) FIXME: A computed property name must be of type 'string',... Remove this comment to see the full error message
            [(StandardContactFields[field.fieldId] || {}).regApiPath]:
              // @ts-expect-error ts-migrate(2538) FIXME: Type 'string[]' cannot be used as an index type.
              existingContactInfo[(StandardContactFields[field.fieldId] || {}).regApiPath]
          }
    );
    const temp = copiedContactInfo.reduce(
      (copiedInfo, info) => (info.customFields ? merge(copiedInfo, info) : Object.assign(copiedInfo, info)),
      {}
    );
    eventRegistration.attendee.personalInformation = temp;

    await dispatch(clearOutOfSyncUserSessionFieldsAndUrlQuery());
    // Reversing the order so that inviteeId is set to undefined for sourceId to be shown as editable.
    dispatch({
      type: CLEAR_REG_CART_INFERRED_FIELDS,
      payload: {
        regCart: setIn(
          getRegCart(state),
          ['eventRegistrations', eventRegistration.eventRegistrationId],
          eventRegistration
        )
      }
    });
  };
}
