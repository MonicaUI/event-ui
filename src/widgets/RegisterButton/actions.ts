import { routeToPage, redirectToPage } from '../../redux/pathInfo';
import { logoutRegistrant } from '../../redux/registrantLogin/actions';
import { openStartNewRegistrationDialogDuringRegistration } from '../../dialogs';
import * as currentRegistrant from '../../redux/selectors/currentRegistrant';
import { DECLINE, GUEST_REGISTRATION, REGISTRATION } from '../../redux/website/registrationProcesses';

/**
 * @param props.config.text Registration dialog text
 * @param props.style Styling for registration dialog
 * @param props.classes Classnames for registration dialog
 * Based on user state and current page
 * Dispatch actions either to redirect to registration start,
 * logout and redirect to 'register' page, or prompt with new registration dialog.
 * RegisterNow widget can exist in each of the conditions, all actions are available
 * RegisterAnother widget can only exists on post-reg pages,
 * openStartNewRegistrationDialog is the only available action
 */
export const promptToStartNewRegistration = (props: $TSFixMe) => {
  return async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const state = getState();
    const loggedIn = currentRegistrant.isLoggedIn(state);
    const registrationPage = REGISTRATION.isTypeOfCurrentPage(state) || GUEST_REGISTRATION.isTypeOfCurrentPage(state);
    const registrationDeclinePage = DECLINE.isTypeOfCurrentPage(state);
    const titleText = props.config.text.htmlContent
      ? props.config.text.htmlContent.replace(/<[^>]*>/g, '')
      : props.config.text;
    if (!loggedIn && !registrationPage && !registrationDeclinePage) {
      return dispatch(routeToPage('register'));
    }
    if (registrationDeclinePage) {
      await dispatch(logoutRegistrant());
      return dispatch(redirectToPage('register'));
    }
    return dispatch(
      openStartNewRegistrationDialogDuringRegistration({
        title: titleText,
        classes: { ...props.classes },
        style: props.style,
        startWithDefaultRegPath: props.startWithDefaultRegPath || false
      })
    );
  };
};
