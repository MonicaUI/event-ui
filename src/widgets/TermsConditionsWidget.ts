import { connect } from 'react-redux';
import TermsConditionsWidget from 'event-widgets/lib/TermsConditions/TermsConditionsWidget';
import { setEventRegistrationFieldValue } from '../redux/registrationForm/regCart/actions';
import {
  isRegistrationModification,
  getAttendeeRegistrationStatus,
  getEventRegistrationId,
  isTermsConditionsAccepted
} from '../redux/selectors/currentRegistrant';
import { getRegCart } from '../redux/selectors/shared';

export function setTermsAndConditionsAcceptance(eventRegistrationId: $TSFixMe, value: $TSFixMe): $TSFixMe {
  return setEventRegistrationFieldValue(eventRegistrationId, ['attendee', 'termsAndConditionsAccepted'], value);
}

/**
 * Data wrapper for the Terms & Conditions widget.
 */
export default connect(
  (state: $TSFixMe) => {
    const eventRegistrationId = getEventRegistrationId(state);
    const termsConditionsAcceptance = isTermsConditionsAccepted(state, eventRegistrationId);
    const isRegMod = isRegistrationModification(state);
    const isRegistered = getAttendeeRegistrationStatus(state, eventRegistrationId) != null;
    const isRegCartRegModThenTransient =
      (getRegCart(state) || {}).status === 'TRANSIENT' && state.regCartStatus?.lastSavedRegCart?.regMod;
    const hidden = (isRegMod && isRegistered) || isRegCartRegModThenTransient;

    return {
      required: !hidden,
      hidden,
      termsConditionsAcceptance,
      eventRegistrationId
    };
  },
  { onSetTermsAndConditionsAcceptance: setTermsAndConditionsAcceptance },
  (stateProps: $TSFixMe, dispatchProps: $TSFixMe, ownProps: $TSFixMe) => {
    return {
      ...ownProps,
      ...stateProps,
      onSetTermsAndConditionsAcceptance: dispatchProps.onSetTermsAndConditionsAcceptance.bind(
        null,
        stateProps.eventRegistrationId
      )
    };
  }
)(TermsConditionsWidget);
