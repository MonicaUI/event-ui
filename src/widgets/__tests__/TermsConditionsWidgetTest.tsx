import React from 'react';
import { mount } from 'enzyme';
import TermsConditionsWidget from '../TermsConditionsWidget';
import { setEventRegistrationFieldValue } from '../../redux/registrationForm/regCart/actions';
jest.mock('../../redux/registrationForm/regCart/actions');

const eventRegistrationId = 'eventRegistrationId1';
let isRegMode = false;
function getState() {
  return {
    event: {
      timezone: 2
    },
    timezones: {},
    registrationForm: {
      currentEventRegistrationId: [eventRegistrationId],
      regCart: {
        regCartId: 'regCartId',
        regMod: isRegMode,
        eventRegistrations: {
          [eventRegistrationId]: {
            eventRegistrationId,
            attendee: {
              termsConditionsAccepted: false,
              personalInformation: {}
            },
            registrationStatus: 'REGISTERED'
          }
        }
      }
    }
  };
}
async function dispatch(action) {
  if (typeof action === 'function') {
    await action(dispatch, getState);
  }
}

const subscribe = () => {};
const props = {
  store: { dispatch, getState, subscribe },
  translate: (res, opts) => (opts ? `${res}:${JSON.stringify(opts)}` : res),
  style: {
    palette: {
      primary: ''
    }
  },
  config: {
    headerText: '',
    instructionalText: '',
    termsConditionsText: ''
  }
};

describe('TermsConditionsWidget', () => {
  test('should render when not hidden', () => {
    const widget = mount(<TermsConditionsWidget {...props} />);
    expect(widget).toMatchSnapshot();
    widget.find('[id="agreementText_0"]').simulate('change');
    expect(setEventRegistrationFieldValue).toHaveBeenCalledWith(
      eventRegistrationId,
      ['attendee', 'termsAndConditionsAccepted'],
      true
    );
  });

  test('should not render when hidden', () => {
    isRegMode = true;
    const updatedProps = { ...props };
    const widget = mount(<TermsConditionsWidget {...updatedProps} />);
    expect(widget).toMatchSnapshot();
  });

  test('should not render when previous reg cart is reg mod,and current reg cart is in TRANSIENT', () => {
    const state = getState();
    function getLocalState() {
      return {
        ...state,
        registrationForm: {
          ...state.registrationForm,
          regCart: {
            ...state.registrationForm.regCart,
            regMod: false,
            status: 'TRANSIENT'
          }
        },
        regCartStatus: {
          lastSavedRegCart: {
            regMod: true
          }
        }
      };
    }
    const updatedProps = {
      ...props,
      store: { dispatch, getState: getLocalState, subscribe }
    };
    const widget = mount(<TermsConditionsWidget {...updatedProps} />);
    expect(widget).toMatchSnapshot();
  });
});
