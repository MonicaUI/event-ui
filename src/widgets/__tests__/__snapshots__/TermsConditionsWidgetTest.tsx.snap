// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TermsConditionsWidget should not render when hidden 1`] = `
Array [
  null,
]
`;

exports[`TermsConditionsWidget should not render when previous reg cart is reg mod,and current reg cart is in TRANSIENT 1`] = `
Array [
  null,
]
`;

exports[`TermsConditionsWidget should render when not hidden 1`] = `
Array [
  <div
    class=""
  >
    <div
      class=""
      data-cvent-id="termsConditions"
      style="flex: 1 1 auto;"
    >
      <h2
        class=""
        data-cvent-id="termsConditions-title"
      />
      <div
        class=""
        data-cvent-id="termsConditions-text"
      />
      <div
        class=" container"
      >
        <div
          class=""
          data-cvent-id="termsConditions-terms"
        />
        <fieldset
          class=""
        >
          <legend
            class=""
            for="agreementText"
          >
            <span
              class=""
            >
              <span
                data-cvent-id="label"
              />
            </span>
          </legend>
          <div
            class=""
          >
            <ul
              class=""
              data-cvent-id="checkbox"
              id="agreementText"
            >
              <li
                data-cvent-id="option-agreementText_0"
              >
                <input
                  class=""
                  id="agreementText_0"
                  name="agreementText"
                  type="checkbox"
                  value="0"
                />
                <label
                  for="agreementText_0"
                >
                  <span>
                    <span
                      class=""
                    >
                       * 
                    </span>
                  </span>
                </label>
              </li>
            </ul>
          </div>
        </fieldset>
      </div>
    </div>
  </div>,
]
`;
